jabgui/src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java

```diff
@@ -68,29 +74,134 @@ private AutoSetFileLinksUtil(List<Path> directories, ExternalApplicationsPrefere
-                LOGGER.error("Problem finding files", e);
-                associatedNotLinkedFiles = Set.of();
-            }
+            doLinkAssociateFiles(entry, onAddLinkedFile, result);
+        }
+        return result;
+    }
 
-            for (LinkedFile associateNotLinkedFile : associatedNotLinkedFiles) {
-                // store undo information
-                onAddLinkedFile.accept(associateNotLinkedFile, entry);
-            }
+    /**
+     * Source of associated not linked files:
+     * Part A. match file name with CitationKey, (START, EXACT, REGEX) configured by user
+     * Part B. match file name with broken linked file names, currently silently happen
+     *
+     * The auto-link process:
+     * Prolog: we only consider the file with a unique name
+     *         if a file's name is found multiple times in Part A, we do not consider it in Step 1
+     *         if a file's name is found multiple times in Part B, we do not consider it in Step 2
+     * Step 1. try to auto link each broken linked file with Part A at first. For unlinked files left in Part A, if no
+     *         linked file with same name exists, add them
+     *         why `add`: Part A are found mainly by CitationKey, which has strong connection to the entry, so we are
+     *                    confident that we should add them automatically
+     * Step 2. try to auto link each broken linked file with Part B.
+     *         how about `unlinked files left in Part B`: there should be no files left as they are found based on broken
+     *                                                    linked files and are used to fix them. `files left` hints a bug
+     *         one trick: we accumulate Part B after Step 1, so Part A does not affect those broken linked files we use
+     *                    to query Part B
+     */
+    private void doLinkAssociateFiles(BibEntry entry, BiConsumer<List<LinkedFile>, BibEntry> onAddLinkedFile, LinkFilesResult result) {
+        boolean entryUpdated = false;
+        // Step 1: try matched files based on CitationKey configured by user
+        Map<String, LinkedFile> files = getAssociatedFiles(entry, result, preConfiguredFileFinder);
+        List<LinkedFile> newLinkedFiles = autoLinkBrokenLinkedFiles(entry.getFiles(), files);
+        if (filesChanged(newLinkedFiles, entry.getFiles())) {
+            entryUpdated = true;
+            onAddLinkedFile.accept(newLinkedFiles, entry);
+        }
+        // Add left unlinked files as new linked files
+        Set<String> linkedFileNames = entry.getFiles().stream().map(LinkedFile::getLink)
+                                           .map(FileUtil::getBaseName).collect(Collectors.toSet());
+        List<LinkedFile> addedFiles = files.keySet().stream()
+                                           .filter(name -> !linkedFileNames.contains(name))
+                                           .map(files::get)
+                                           .toList();
```

Comment:
We can save a `files` map lookup per iteration by taking the key value pair together:

```suggestion
        List<LinkedFile> addedFiles = files.entrySet().stream()
                                           .filter(entry -> !linkedFileNames.contains(entry.getKey()))
                                           .map(Map.Entry::getValue)
                                           .toList();
```

Reply:
fixed. Thx for the `Optional::stream` trick